pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,459 @@
+			log.Fatalf(r.ctx, "numRegisteredReplica(%d) > numAllowed(%d)", numRegisteredReplica, numAllowed)
+		}
+	}
+
+	// Only increment the number of registered traces if the numRegistered
+	// hasn't changed. In the case of an ABA update, it does not break the
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available.
+	registerSucceeded := r.numRegisteredStore.CompareAndSwap(numRegisteredStore, numRegisteredStore+1)
+	if registerSucceeded {
+		// Add one unconditionally to the replica count.
+		r.numRegisteredReplica.Add(1)
+	}
+	// Note we can't assert numRegisteredStore <= numAllowed because if the
+	// setting is changed it can be temporarily violated on other replicas.
+	return registerSucceeded
+}
+
+// storeEntry stores the trace and returns the newly created value.
+func (r *RaftTracer) storeEntry(
+	te kvserverpb.TracedEntry, propCtx context.Context, span *tracing.Span,
+) *traceValue {
+	tv := traceValue{
+		traced: te,
+	}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+
+	tv.mu.propCtx = propCtx
+	tv.mu.propSpan = span
+	tv.mu.seenMsgAppResp = make(map[raftpb.PeerID]bool)
+	tv.mu.seenMsgStorageAppendResp = make(map[raftpb.PeerID]bool)
+	r.m.Store(te.Index, &tv)
+	return &tv
+}
+
+// RegisterRemote registers a remote trace. This is called when we receive a
+// raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(te kvserverpb.TracedEntry) {
+	if !r.reserveSpace() {
+		return
+	}
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	tv := r.storeEntry(te, nil, nil)
+	tv.logf(1, "registering remote trace %s", tv)
+}
+
+// MaybeRegister is called on an entry that is about to be proposed. This will
+// begin logging all subsequent updates to this entry. It will return true if
+// the registration succeeded.
+func (r *RaftTracer) MaybeRegister(ctx context.Context, ent raftpb.Entry) bool {
+	// If the index is nil, then we can't trace this entry. This can happen if
+	// there is a leader/leaseholder spilt. We don't have an easy way to handle
+	// this today, so don't attempt to trace it.
+	if ent.Index == 0 {
+		log.VEventf(ctx, 2, "skip registering raft proposal without index: %v", ent)
+		return false
+	}
+
+	// Only register if there is a trace in the context and it is set to verbose
+	// logging.
+	span := tracing.SpanFromContext(ctx)
+	if span == nil || span.RecordingType() != tracingpb.RecordingVerbose {
+		return false
+	}
+
+	// This must be the last conditional. If this returns true we must call
+	// storeEntry to not leak a registered permit.
+	if !r.reserveSpace() {
+		log.VEvent(ctx, 2, "too many active raft traces, skipping")
+		return false
+	}
+
+	ctx, span = r.tracer.StartSpanCtx(ctx, "raft trace",
+		tracing.WithParent(span), tracing.WithFollowsFrom())
+
+	tv := r.storeEntry(
```

Comment:
nit: convention is to place context first. Suggest something like:

```golang
tv := r.storeEntry(ctx, span, kvserverpb.TracedEntry {
	// ...
})
```

Reply:
Good point, reordered.